server/pipeline.go

```diff
@@ -55,6 +55,35 @@ func (s *Server) composePipelineRelationship(ctx context.Context, pipeline *api.
 	return nil
 }
 
+func (s *Server) composePipelineRelationshipValidateOnly(ctx context.Context, pipeline *api.Pipeline) error {
+	var err error
+
+	pipeline.Creator, err = s.composePrincipalByID(ctx, pipeline.CreatorID)
+	if err != nil {
+		return err
+	}
+
+	pipeline.Updater, err = s.composePrincipalByID(ctx, pipeline.UpdaterID)
+	if err != nil {
+		return err
+	}
+
+	if pipeline.StageList == nil {
+		pipeline.StageList, err = s.composeStageListByPipelineID(ctx, pipeline.ID)
```

Comment:
In validate only case, pipeline.StageList is not nil so that we can remove this block.